packages/vite-plugin-cloudflare/src/node-js-compat.ts

```diff
@@ -1,18 +1,124 @@
+import { defineEnv } from "unenv";
+import { resolvePluginConfig } from "./plugin-config";
+import type {
+	PluginConfig,
+	ResolvedPluginConfig,
+	WorkerConfig,
+} from "./plugin-config";
+import type { Plugin } from "vite";
+
+/**
+ * A Vite plugin that can provide Node.js compatibility support for Vite Environments that are hosted in Cloudflare Workers.
+ */
+export function nodejsCompatPlugin(pluginConfig: PluginConfig): Plugin {
+	let resolvedPluginConfig: ResolvedPluginConfig;
+	return {
+		name: "vite-plugin-nodejs-compat",
+		config(userConfig, env) {
+			// Capture the configuration of the Cloudflare Workers for use in other hooks, below.
+			resolvedPluginConfig = resolvePluginConfig(pluginConfig, userConfig, env);
+
+			// Configure Vite with the Node.js polyfill aliases
+			// We have to do this across the whole Vite config because it is not possible to do it per Environment.
+			return {
+				resolve: {
+					alias: getNodeCompatAliases(),
+				},
+			};
+		},
+		configEnvironment(environmentName) {
+			if (resolvedPluginConfig.type === "assets-only") {
+				return;
+			}
+
+			const workerConfig = resolvedPluginConfig.workers[environmentName];
+			if (!workerConfig) {
+				return;
+			}
+
+			if (!isNodeCompat(workerConfig)) {
+				return;
+			}
+
+			// Ignore the Node.js external modules when building.
+			return {
+				build: {
+					rollupOptions: {
+						external: getNodeCompatExternals(),
+					},
+				},
+			};
+		},
+		async resolveId(source) {
+			if (resolvedPluginConfig.type === "assets-only") {
+				return;
+			}
+
+			const workerConfig = resolvedPluginConfig.workers[this.environment.name];
+			if (!workerConfig) {
+				return;
+			}
+
+			const unresolvedAlias = dealiasVirtualNodeJSImports(source, workerConfig);
+			if (!unresolvedAlias) {
+				return;
+			}
+
+			const resolvedAlias = await this.resolve(
+				unresolvedAlias,
+				import.meta.url
+			);
+			if (!resolvedAlias) {
+				return;
+			}
+
+			if (this.environment.mode === "dev" && this.environment.depsOptimizer) {
+				// Make sure the dependency optimizer is aware of this aliased import
+				this.environment.depsOptimizer.registerMissingImport(
+					unresolvedAlias,
+					resolvedAlias.id
+				);
+			}
+
+			return resolvedAlias;
+		},
+		async transform(code, id) {
+			if (resolvedPluginConfig.type === "assets-only") {
+				return;
+			}
+
+			const workerConfig = resolvedPluginConfig.workers[this.environment.name];
+			if (!workerConfig) {
+				return;
+			}
+
+			if (!isNodeCompat(workerConfig)) {
+				return;
+			}
+
+			const resolvedId = await this.resolve(workerConfig.main);
+			if (id === resolvedId?.id) {
+				return injectGlobalCode(id, code);
+			}
+		},
+	};
+}
+
+const { env } = defineEnv({
+	nodeCompat: true,
+	presets: [cloudflare],
+});
 
-const require = createRequire(import.meta.url);
-const preset = unenv.env(unenv.nodeless, unenv.cloudflare);
 const CLOUDFLARE_VIRTUAL_PREFIX = "\0cloudflare-";
```

Comment:
Maybe let's adopt this convention for private module prefixes to reduce the risk of collisions:

```suggestion
const CLOUDFLARE_VIRTUAL_PREFIX = "\0__CLOUDFLARE_NODEJS_COMPAT__";
```

Reply:
cb4388afa97d8a7132c7dc5b7203558d47acc668